rllib/algorithms/algorithm.py

```diff
@@ -941,6 +943,199 @@ def duration_fn(num_units_done):
+        )
+
+        if self.config["custom_eval_function"]:
+            raise ValueError(
+                "`custom_eval_function` not supported in combination "
+                "with `evaluation_with_async_requests=True` config setting!"
+            )
+        if (
+            self.evaluation_workers is None
+            and (
+                self.workers.local_worker().input_reader is None
+                or self.config["evaluation_num_workers"] == 0
+            )
+        ):
+            raise ValueError(
+                "Local evaluation OR evaluation without input reader OR evaluation "
+                "with only a local eval worker not supported in combination "
+                "with `evaluation_with_async_requests=True` config setting!"
+            )
+
+        agent_steps_this_iter = 0
+        env_steps_this_iter = 0
+
+        logger.info(f"Evaluating current policy for {duration} {unit}.")
+
+        all_batches = []
+
+        # Default done-function returns True, whenever num episodes
+        # have been completed.
+        if duration_fn is None:
+            def duration_fn(num_units_done):
+                return duration - num_units_done
+
+        def remote_fn(worker, w_ref, w_seq_no):
+            worker.set_weights(weights=w_ref)
+            batch = worker.sample()
+            metrics = worker.get_metrics()
+            return batch, metrics, w_seq_no
+
+        rollout_metrics = []
+
+        # How many episodes have we run (across all eval workers)?
+        num_units_done = 0
+        round_ = 0
+        time_started = time.time()
+        timed_out = True
+
+        while time.time() - time_started < self.config["evaluation_sample_timeout_s"]:
+            units_left_to_do = duration_fn(num_units_done)
+            if units_left_to_do <= 0:
+                timed_out = False
+                break
+
+            round_ += 1
+            # Use the AsyncRequestsManager to get ready evaluation results and
+            # metrics.
+            self._evaluation_async_req_manager.call_on_all_available(
+                remote_fn=remote_fn,
+                fn_args=[weights_ref, self._evaluation_weights_seq_number],
+            )
+            ready_requests = self._evaluation_async_req_manager.get_ready()
+
+            batches = []
+            for i, requests in enumerate(ready_requests.values()):
+                for req in requests:
+                    batch, metrics, seq_no = req
+                    # Ignore results, if the weights seq-number does not match (is from
+                    # a previous evaluation step) OR if we have already reached the
+                    # configured duration (e.g. number of episodes to evaluate for).
+                    if (
+                        seq_no == self._evaluation_weights_seq_number
+                        and (
+                            i * (1 if unit == "episodes" else rollout_len * num_envs)
+                            < units_left_to_do
```

Comment:
do we really care about this? usually folks don't complain if we have a bit more data than intended? :)

Reply:
"usually" Yeah, but this is cleaner, no? :D